pkg/registrars/registration/ampcache-registrar.go

```diff
@@ -0,0 +1,297 @@
+	if err != nil {
+		return regResp, err
+	}
+	resp, err := r.client.Do(req)
+	if err != nil {
+		logger.Warnf("%v failed to do HTTP request to registration endpoint %s: %v", r.endpoint, err)
+		return nil, err
+	}
+	logger.Printf("Response: %v", resp)
+	defer resp.Body.Close()
+
+	logger.Printf("AMP cache rendezvous response: %s", resp.Status)
+
+	// Check that the HTTP request returned a success code
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		// logger.Warnf("got non-success response code %d from registration endpoint %v", resp.StatusCode, r.endpoint)
+		return nil, fmt.Errorf("non-success response code %d on %s", resp.StatusCode, r.endpoint)
+	}
+
+	if _, err := resp.Location(); err == nil {
+		// The Google AMP Cache may return a "silent redirect" with
+		// status 200, a Location header set, and a JavaScript redirect
+		// in the body. The redirect points directly at the origin
+		// server for the request (bypassing the AMP cache). We do not
+		// follow redirects nor execute JavaScript, but in any case we
+		// cannot extract information from this response and can only
+		// treat it as an error.
+		return nil, fmt.Errorf("non-success silent redirect %d on %s", resp.StatusCode, r.endpoint)
+	}
+	lr := io.LimitReader(resp.Body, readLimit+1)
+	logger.Printf("Returned lr: %s", lr)
+	dec, err := amp.NewArmorDecoder(lr)
+	if err != nil {
+		return nil, fmt.Errorf("Armor Decoder failed with error: %d", err)
+	}
+	bodyBytes, err := io.ReadAll(dec)
+	if err != nil {
+		return nil, fmt.Errorf("ReadAll failed with error: %d", err)
+	}
+	if lr.(*io.LimitedReader).N == 0 {
+		// We hit readLimit while decoding AMP armor, that's an error.
+		return nil, fmt.Errorf("Hit readLimit while decoding AMP armor: %d", io.ErrUnexpectedEOF)
+	}
+	// Unmarshal response body into Registration Response protobuf
+	if err = proto.Unmarshal(bodyBytes, regResp); err != nil {
+
+		return regResp, fmt.Errorf("Error in storing Registration Response protobuf: %v", err)
+	}
+	return regResp, nil
+}
+
+func (r *AMPCacheRegistrar) executeRequestCommon(ctx context.Context, payload []byte, logger logrus.FieldLogger) (*http.Request, error) {
+	// Make new HTTP request with given context, registrar, and paylaod
+	logger.Println("Registering via AMP cache rendezvous...")
+	logger.Println("Station URL:", r.endpoint)
+	logger.Println("AMP cache URL:", r.ampCacheURL)
+
+	endpointURL, err := url.Parse(r.endpoint)
+	if err != nil {
+		logger.Warnf("failed to parse endpoint url")
+	}
+	reqURL := endpointURL.ResolveReference(&url.URL{
+		Path: endpointURL.Path + "/" + amp.EncodePath(payload),
+	})
+
+	// Rewrite reqURL to its AMP cache version.
+	ampURL, err := url.Parse(r.ampCacheURL)
+	if err != nil {
+		logger.Warnf("failed to parse ampcache url")
+	}
+	reqURL, err = amp.CacheURL(reqURL, ampURL, "c")
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", reqURL.String(), bytes.NewReader(payload))
```

Comment:
We don't need to send the payload in the body of the GET request anymore because the AMP cache server will ignore and remove it anyway before forwarding the request to the registration server. We can remove this `bytes.NewReader(payload)` call.